### PR TITLE
Revert "Revert "Ban the cache manager from creating unwanted entries""

### PR DESCRIPTION
**Pull request by Simply007**
Please continue with the pull request:
* https://github.com/Kentico/cloud-boilerplate-net/pull/56

### CODE REVIEW

**Comment by JanLenoch** [comment]
In response to https://github.com/Kentico/cloud-boilerplate-net/pull/56#issuecomment-409311762:

I'll try to list pros and cons of such approach.
Pros:
* Empty listing responses would be cached.

Cons:
* We would have to build a logic to inspect all listing responses upon each webhook calls that would select all empty ones.
* Even such solution would be sub-optimal. Given that all empty listing responses would be smashed upon such webhook call, the advantage listed above would vanish.
* An optimal solution would be to carefully pick only those empty listing responses that *may* contain that content item (or other kind of KC data). The word *may* is emphasized because the cache manager could only guess if a particular cache entry should be invalidated. I think we don't want such an overly intelligent cache management, just for sake of holding empty listing responses.

File: src/content/CloudBoilerplateNet/Services/ReactiveCacheManager.cs [hunk 1]
```diff
@@ -93,14 +95,17 @@ public async Task<T> GetOrCreateAsync<T>(IEnumerable<string> identifierTokens, F
                     // If it doesn't exist, get it via valueFactory.
                     T response = await valueFactory();
 
-                    // Create it in a background thread.
-                    if (createCacheEntriesInBackground)
+                    if (productionValidator(response))
```
Comment (petrsvihlik) [inline src/content/CloudBoilerplateNet/Services/ReactiveCacheManager.cs]: I'd rename `productionValidator` to something like `skipCacheDelegate`. (And adjust/revert the actual conditions accordingly.)

File: src/content/CloudBoilerplateNet/Services/CachedDeliveryClient.cs [hunk 2]
```diff
@@ -287,7 +359,13 @@ public async Task<TaxonomyGroup> GetTaxonomyAsync(string codename)
         {
             var identifierTokens = new List<string> { KenticoCloudCacheHelper.TAXONOMY_GROUP_SINGLE_IDENTIFIER, codename };
 
-            return await CacheManager.GetOrCreateAsync(identifierTokens, () => DeliveryClient.GetTaxonomyAsync(codename), GetTaxonomySingleDependency, ProjectOptions.CreateCacheEntriesInBackground);
+            // The "productionValidator" delegate does not have to do anything since a DeliveryException is thrown instead.
```
Comment (petrsvihlik) [inline src/content/CloudBoilerplateNet/Services/CachedDeliveryClient.cs]: I'd get rid of all those comments. They are specific for one case and they wouldn't be understood by newcomers.

File: src/content/CloudBoilerplateNet/Services/CachedDeliveryClient.cs [hunk 3]
```diff
@@ -287,7 +359,13 @@ public async Task<TaxonomyGroup> GetTaxonomyAsync(string codename)
...
+            return await CacheManager.GetOrCreateAsync(
+                identifierTokens, 
+                () => DeliveryClient.GetTaxonomyAsync(codename), 
+                response => true,
```
Comment (petrsvihlik) [inline src/content/CloudBoilerplateNet/Services/CachedDeliveryClient.cs]: I'd replace the `true` value with e.g. `response != null`. We know that currently the Delivery SDK won't return anything and will throw an exception instead. However, in the future that might change and this code will remain valid and understandable (even without the comment above).

**Pull request merged**